File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxx=prices[-1]
        ans=0
        for r in range(len(prices)-1,-1,-1):
            ans=max(ans,maxx-prices[r])
            maxx=max(maxx,prices[r])
        return ans

File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int firstBuy=-1000000;
        int firstSell=0;
        for(int p:prices){
            firstBuy=Math.max(firstBuy,-p);
            firstSell=Math.max(firstSell,p+firstBuy);
        }
        return firstSell;
    }
}